### PR TITLE
[CoreWorker] Partially address Ray child process leaks by killing all child processes in the CoreWorker shutdown sequence.

### LINKED ISSUE #31451: [core] Multi-process(?) / GPU processes do not seem to be freed after ctrl+c on cluster
### What happened + What you expected to happen

This has been discussed in #30414 at the end, but it seems more approriate to start a new issue because it is probably unrelated.

I run a cluster with a head node, and worker nodes. The head node gets a CLI `ray start ...options`, the worker nodes also do a CLI `ray start .. options` and connect to the head. All the worker nodes have GPUs.

On the head node I then run a tune script with GPU request. After ctrl+c, the GPU does not seem to be freed on the worker nodes, and IDLE or TRAIN processes remain that block the GPU memory. Only a full `ray stop` kills all the processes. 

I tested this also in a much more simple setup (just a head node with a GPU), run the tune script, do ctrl+c, and the memory on the GPU remains blocked and ray:TRAIN processes remain (see #30414).

_**EDIT: The issue was found to be related to `num_workers>0` in pytorch DataLoader, which leaves extra ray processes open after ctrl+c.**_. Related: https://github.com/ray-project/ray_lightning/issues/87, https://github.com/pytorch/pytorch/issues/66482

~~_**EDIT 2: I could solve the issue by using 515.xxx NVIDIA drivers (but only for the main node), but for 470.xxx and/or the worker nodes the issue seems to remain.**_~~ 

**EDIT 3: The issue persists, irregardless of driver version**

### Versions / Dependencies

Ray 2.2.0
Pytorch 1.12.1
NVIDIA driver 470.xxx

### Reproduction script

Something like
```
if(args.resume):
  tuner = Tuner.restore(
      path=os.path.join(results_dir, "test")
  )
  
  tuner.fit()
else:
  trainable=tune.with_resources(trainable, resources={"cpu": 4, "gpu": 1})
  
  failure_config=FailureConfig(max_failures=-1)
  
  run_config=RunConfig(name="test",
                       local_dir=results_dir,
                       failure_config=failure_config,
                       log_to_file=True
                       )
  
  tune_config=TuneConfig(num_samples=1,
                         reuse_actors=False
                         )
  
  tuner=Tuner(new_trainable, run_config=run_config, tune_config=tune_config, param_space=cfg)
  
  tuner.fit()
```

Any tune job that runs on a node that is previously started with `ray start --num_cpus=4 --num_gpus=1`.

### Issue Severity

Medium: It is a significant difficulty but I can work around it.

### PR DESCRIPTION
**Pull request by cadedaniel**
### TL;DR
We kill all child processes when a Ray worker process exits. This addresses process leaks that caused GPU OOM errors in https://github.com/ray-project/ray/issues/31451. There is some risk to this PR, particularly if Ray users rely on Ray's existing behavior of leaking processes. We don't know of any such user, but we provide a new flag `RAY_kill_child_processes_on_worker_exit` to provide a workaround in case someone is impacted.

### Motivation
In https://github.com/ray-project/ray/issues/31451 we find that Ray workers can "leak" child processes in certain cases. This was manifest as a GPU OOM because a leaked worker child process kept a CUDA context alive, which kept the reserved GPU memory allocated. Future Ray tasks/actors could no longer use the full GPU memory until users ran `ray stop`. This is surprising to users because they don't expect Ray to leave their cluster in a bad state.

Thus, Ray needs some way to ensure that reasonably well-behaved code doesn't cause resource starvation due to leaked processes.

#### Minimal repro (from @matthewdeng https://github.com/ray-project/ray/issues/31451#issuecomment-1372797046)

After `Ctrl+C`, this will leak the child process. If that child process were to hold GPU memory, future workloads on the same machine would not be able to use the memory.
```python
import ray
import time
from multiprocessing import Process

@ray.remote
class MyActor:

    def run(self):
       p = Process(target = lambda: time.sleep(1000), daemon=True)
       p.start()
       p.join()

actor = MyActor.remote()
ray.get(actor.run.remote())
```

#### Root cause 1
This bug is caused by a bug in the cleanup flow for worker processes, particularly actors. We set [force_kill to true](https://github.com/ray-project/ray/blob/bb24c16b63578c7ba2cc2f250ef9aad2be2278cc/src/ray/gcs/gcs_server/gcs_actor_manager.cc#L749-L752) when killing actors that go out-of-scope. This effectuates an immediate termination of the worker process, such that `atexit` handlers and destructors are not guaranteed to be called. In particular, `multiprocessing.Process` installs an `atexit` handler which terminates all known daemon child processes.

Because of this bug in our cleanup logic, `multiprocessing.Process` is unable to clean up as it would in a vanilla Python interpreter, and we are left with a leak. Notably, this bug could strike any case where destructors/atexit handlers clean up machine state. 

We should fix this bug, filed here https://github.com/ray-project/ray/issues/34124.

#### Root cause 2
Even if we fix the root cause of the leak, it will still be possible for buggy code to leak processes which affect the state of the cluster. For example, someone could intentionally double-fork and get a grandchild process reparented to the init process. Malicious intent is not required; bugs in user code or libraries could achieve the same thing.

This is a bad experience for Ray users expecting a native Python experience because it is relatively very difficult to clean up an entire cluster compared to cleaning up leaked state on a single machine as one would with a buggy vanilla-Python program. Thus, Ray should have some isolation guarantees so that buggy tasks or actors do not break a cluster to the point of requiring manual intervention.

#### Root Cause 1
We should improve the cleanup flow for actors so that destructors and atexit handlers can run as they do in a vanilla Python program. We attempted this fix before the 2.4 release but feel that it is too risky to merge at this late stage in the release. Following up here  https://github.com/ray-project/ray/issues/34124. 

#### Root Cause 2
See [this thread](https://github.com/ray-project/ray/pull/33976#discussion_r1154846845) for discussion on potential solutions for Root Cause 2.

The cleanest solution is to configure the raylet to be a Linux child subreaper for all Ray worker processes and their descendants. The raylet would then periodically scan its immediate children and kill any it does not recognize as a worker.

A partial solution to this problem is to have the CoreWorker kill its children during the shutdown sequence. This handles the user-reported bug where the process leaks during a Ctrl+C, but does not handle leaks originating from worker crashes.

We decided to go with the partial solution, mainly because due to timing around the release and us not discovering Root Cause 1 or the PR_SET_CHILD_SUBREAPER until late in the 2.4 release process. We will follow up with a better solution to Root Cause 2, see https://github.com/ray-project/ray/issues/34125 for tracking.

### Testing
I reproduce the minimal process leak case provided above and verify that all child processes (whether they're started by actors or tasks) are terminated upon sending Ctrl+C to the driver.

### Closes
Closes https://github.com/ray-project/ray/issues/31451

Remaining items:
- [x] Automated testing
- [x] Clean up names / add comments

### CODE REVIEW

File: src/ray/core_worker/core_worker.cc [hunk 1]
```diff
@@ -704,6 +704,35 @@ void CoreWorker::Disconnect(
   }
 }
 
+void CoreWorker::KillLeakedProcs() {
```
Comment (cadedaniel) [inline src/ray/core_worker/core_worker.cc:707]: I call this in the CoreWorker. There are cases where we will still leak processes, like if the worker process segfaults. To handle this, we should keep track of child processes of the workers in the raylet. I am intentionally not doing that now to 1) get validation on this approach and 2) get a faster fix in because that kind of resource tracking is a larger change.

Note that even if we track child processes in the raylet, we'll still need code here in case the raylet dies.

Comment (ericl) [reply src/ray/core_worker/core_worker.cc:707]: I think this is one of those cases where you do want an e2e correct solution on the first try. I don't think it makes sense to invest effort on a solution that isn't fully correct to this extent..

Comment (ericl) [reply src/ray/core_worker/core_worker.cc:707]: What was the result of the investigation around the raylet killing the reparented child processes as the process group leader?

Comment (cadedaniel) [reply src/ray/core_worker/core_worker.cc:707]: > What was the result of the investigation around the raylet killing the reparented child processes as the process group leader?

Killing the reparented child processes by process group was not part of the investigation because they are in the raylet's process group (so all workers and any child processes would also get killed).

As for killing the reparented child processes, I assumed that they would be re-parented to the raylet. This was incorrect; I tried it and they are reparented to init (pid 1).

> I think this is one of those cases where you do want an e2e correct solution on the first try. I don't think it makes sense to invest effort on a solution that isn't fully correct to this extent..

The spectrum of solutions I see are as follows:

1. Kill processes by cgroup (best userspace solution).
  * Sureproof cleaning up of accidentally-leaked processes.
  * Significant re-architecture of how Ray creates/manages worker processes.
  * Requires handling Raylet crashes.

2. Kill processes by process-group.
  * Sureproof cleaning up of accidentally-leaked processes.
  * Nontrivial changes in Ray given that Ray currently places all workers in the same process group and uses that for cleanup.
  * Requires handling Raylet crashes.

3. Track worker child processes in the raylet; kill child processes once worker dies.
  * Not sureproof clean up -- will leak processes in certain race conditions, and can kill arbitrary new processes in other race conditions.
  * Straightforward but nontrivial addition to raylet.
  * Requires handling Raylet crashes.

4. Kill worker child process when a worker exits, either from Cancellation (ctrl+C) or on normal teardown.
  * Sureproof clean up of accidentally-leaked processes except in worker segfault.
  * Straightforward and small addition to core worker.
  * The raylet can crash and workers still clean up correctly.

Horable mention: Raylet kills processes that are in the Raylet process group but have ppid 1. It is not guaranteed that ppid==1 for orphaned processes.

I see (2) as the best medium-term solution. I think (4) is the better short-term solution between (3, 4) because it is sureproof under user cancellation or normal teardown, and easy to implement.

Comment (ericl) [reply src/ray/core_worker/core_worker.cc:707]: > Horable mention: Raylet kills processes that are in the Raylet process group but have ppid 1. It is not guaranteed that ppid==1 for orphaned processes.

Why isn't this a candidate solution? It seems to (1) be correct e2e, and (2) pretty straightforward to implement via polling, for example. Even if ppid != 1, we could do it for any unknown ppid.

Comment (ericl) [reply src/ray/core_worker/core_worker.cc:707]: Btw, https://unix.stackexchange.com/questions/149319/new-parent-process-when-the-parent-process-dies

suggests that you can use a `prctl` call to request orphaned children to be reparented to yourself.

Comment (jjyao) [reply src/ray/core_worker/core_worker.cc:707]: Oh, didn't know that we have this. Seems that's what we want.

Comment (cadedaniel) [reply src/ray/core_worker/core_worker.cc:707]: > >Horable mention: Raylet kills processes that are in the Raylet process group but have ppid 1. It is not guaranteed that ppid==1 for orphaned processes.
> 
> Why isn't this a candidate solution? It seems to (1) be correct e2e, and (2) pretty straightforward to implement via polling, for example. Even if ppid != 1, we could do it for any unknown ppid.

I avoided that solution because you have to distinguish between leaked child processes and non-leaked child processes (since at depth>1 child processes will have unknown ppids). And climbing the ppid tree has race conditions which are especially risky if there are many child processes being created.

That said, we could do something like a mark-and-sweep approach to drastically reduce the chance of these race conditions; on the first poll, mark candidate processes for termination, on the second poll confirm the tree structure is the same and then kill the child process.

This still fails when the raylet crashes, since the entire old raylet process group will not be known.

I think this is potentially something we do before (2) above. But even this case will require the core-worker to kill child processes on exit to handle raylet crashes (which is the code in this PR).

Comment (cadedaniel) [reply src/ray/core_worker/core_worker.cc:707]: >Btw, https://unix.stackexchange.com/questions/149319/new-parent-process-when-the-parent-process-dies
>
> suggests that you can use a prctl call to request orphaned children to be reparented to yourself.

Oh, nice find. Seems that we can use this so that all leaked processes get reparented to the raylet. The missing piece then would be to periodically poll children of the raylet and if a process is not recognized (i.e. isn't a known worker), then we assume it is a leaked child process and we kill it.

Can we still merge this coreworker change to handle the raylet crash case, and since we believe it addresses the high-frequency torch dataloader process leak?

Comment (jjyao) [reply src/ray/core_worker/core_worker.cc:707]: I think if raylet and worker crashes the same time, we can still have leaks.

This is my current understanding (might be wrong): with the current linux mechanisms, unless the child process is cooperative (e.g. using PR_SET_PDEATHSIG, listening to pipe), there is always corner case where the child process is leaked.

**Comment by rkooo567** [comment]
Hey guys. I'd like to make a counter proposal to fix this issue here. I'd like to hear your opinions. 

I made a PR (https://github.com/ray-project/ray/pull/34045) that fixes the multi-processing leak issue that @matthewdeng created. It is not ready yet (need more testing to merge), but I verified it fixes the multi-processing leak issue. 

This fix assumes one thing; (potentially controversial); **When it comes to child procs cleanup, we don't need to support more than what Python provides.** For example, daemon processes are not cleaned up when the Python script is killed by SIGKILL, and that's the expected behavior. IMO, we don't need to provide a stronger cleanup guarantee here (since users are aware of the limitation, and most of the libraries should be implemented with this in mind). But this can be discussed.

## Problems

For both Pytorch dataloader and MultiProcessing daemon processes, they are cleaned up from the destructor, `__del__` . `__del__` is called under the proper process termination process. However, it is not called when SIGKILL is received or the process is terminated via _Exit(), which is the implementation detail of `QuickExit` in Ray (which is supposed to be used under "emergency scenario").

The root cause of the issue is that we don't properly exit the process when actors are terminated. We bypass the proper termination path "even under common scenarios". Here's more detail about how actors exit in Ray. 

1. When the actor handle goes out of scope, GCS sends an RPC to the core worker which calls QuickExit -> However, it bypasses the destructor (thus we can have leaks like this).
2. When the job terminates, it sends SIGTERM -> SIGKILL to all owned workers. However, we have bugs in handling SIGTERM for actors (that's something I fixed in my PR). Simply put, if we send a SIGTERM to an actor, it triggers the check failure. Also SIGTERM -> SIGKILL interval is really short (100ms).
3. When `ray.kill` is called, GCS sends an RPC to the core worker to call QuickExit, which has the same problem as (1).

Although `QuickExit` comes in handy sometimes, for these 3 scenarios, users would expect the process to exit properly, calling all of their `__del__` methods.

Also note that since most of the cleanup logic is implemented in the destructor, it means we always have a risk of random "resource leaks" from any Python code although we fix the process leak issues by the proposal in this PR.

# Proposal

My counter-proposal is to make sure we exit actor processes properly. And then we can probably work on child processes cleaning things if necessary (I think this will solve 99% of user issues). I believe it is already the case for tasks (I verified). It is a bug that actor processes don't properly exit.

There are 3 work items;
- We kill actors in the same way in every scenario. Raylet sends SIGTERM. It waits for the grace period (I am thinking 5~10 seconds). And then send a SIGKILL
- We always make Raylet kill the actors. When GCS wants to kill actors, send an RPC to the corresponding raylet.
- We fix the issue the actor processes cannot handle SIGTERM properly. I verified when I just send a SIGTERM to actors, it has a really bad check failure.

The PR I have fixed the 3rd issue. Though I need more testing to make sure this works well.

**Comment by jjyao** [comment]
I think there are several decisions we need to make:

1. When a worker process normally exits, is it ok to have child process leak
2. When a worker process crashes, is it ok to have child process leak
3. When raylet crashes, is it ok to have child process leak

@rkooo567 I think your PR tries to fix 1 and we are trying to fix 2 & 3.

1 is noncontroversial so it's really about 2 & 3.

**Comment by cadedaniel** [comment]
Yeah, @rkooo567 your fix should also go in (nice find) but doesn't address case where child process misbehaves.

But your approach highlights something interesting.. really, we should kill child processes once the task or actor finishes, _not_ when the worker process exits. Otherwise unkilled child processes from a dead task will continue to consume system resources (e.g. GPU memory) since the worker process is alive.

I think fixing this iteratively is a good approach.. not sure if this is an issue people see.

**Comment by cadedaniel** [comment]
After more investigation, I found that Sang's proposal requires a bugfix in actor cleanup behavior so that actor processes are properly cleaned up. The fix is not huge but there is risk if we want to get it in by 2.4.

This PR addresses the process leak reported by users, where processes leak after pressing Ctrl+C. It does not address leaks that come from crashes in worker processes. We'll need to improve upon this in future releases to harden Ray against worker child process leaks.

File: python/ray/tests/test_failure_3.py [hunk 2]
```diff
@@ -347,6 +352,103 @@ def pid(self):
         ray.get(t)
 
 
+def test_no_worker_child_process_leaks(ray_start_cluster, tmp_path):
```
Comment (rkooo567) [inline python/ray/tests/test_failure_3.py:355]: is it possible to add a test with torch data loader?

File: src/ray/core_worker/core_worker.cc [hunk 3]
```diff
@@ -704,6 +704,55 @@ void CoreWorker::Disconnect(
...
+
+  // Enumerating child procs is not supported on this platform.
+  if (!maybe_child_procs) {
+    RAY_LOG(DEBUG) << "Killing leaked procs not supported on this platform.";
+    return;
+  }
+
+  const auto &child_procs = *maybe_child_procs;
+  const auto child_procs_str = absl::StrJoin(child_procs, ",");
+  RAY_LOG(DEBUG) << "Procs to kill: " << child_procs_str;
```
Comment (rkooo567) [inline src/ray/core_worker/core_worker.cc:739]: Can you make this INFO log and comment "Try Killing all child processes of this worker as it exits. Child process pids: 23,45,67..."

File: src/ray/core_worker/core_worker.cc [hunk 4]
```diff
@@ -704,6 +704,55 @@ void CoreWorker::Disconnect(
...
+
+  for (const auto &child_pid : child_procs) {
+    auto maybe_error_code = KillProc(child_pid);
+    RAY_CHECK(maybe_error_code &&
+              "Expected this path to only be called when KillProc is supported.");
+    auto error_code = *maybe_error_code;
+
+    RAY_LOG(DEBUG) << "Kill result for " << child_pid << ": " << error_code.message()
```
Comment (rkooo567) [inline src/ray/core_worker/core_worker.cc:747]: I think this also can be a INFO log because it happens only once upon termination. And it is better making it clear what's happening upon termination.

**Review by rkooo567 — APPROVED** [review]
nits only

File: python/ray/tests/test_failure_3.py (hunk 2 above)
Comment (cadedaniel) [reply python/ray/tests/test_failure_3.py:355]: Sure, let me try today

File: src/ray/util/process.cc [hunk 5]
```diff
@@ -636,6 +636,87 @@ bool IsProcessAlive(pid_t pid) {
...
+
+  // Iterate over all files in the /proc directory, looking for directories.
+  // See `man proc` for information on the directory structure.
+  // Directories with only digits in their name correspond to processes in the process
+  // table. We read in the status of each such process and parse the parent PID. If the
+  // process parent PID is equal to parent_pid, then we add it to the vector to be
+  // returned. Ideally, we use a library for this, but at the time of writing one is not
+  // available in Ray C++.
+
+  std::filesystem::directory_iterator dir(kProcDirectory);
```
Comment (scv119) [inline src/ray/util/process.cc:669]: does this work in docker container too (and is it configurable?)

**Review by scv119 — APPROVED** [review]

File: src/ray/util/process.cc (hunk 5 above)
Comment (cadedaniel) [reply src/ray/util/process.cc:669]: Yes, works in docker container. The [man pages](https://manpages.ubuntu.com/manpages/xenial/man5/proc.5.html) suggest that there is a way you can change it, but I think it's really rare.

> The  proc  filesystem  is  a  pseudo-filesystem which provides an interface to kernel data
       structures.  It is commonly mounted at [/proc](file:///proc).  Most of it is  read-only,  but  some  files
       allow kernel variables to be changed.

File: src/ray/util/process.cc [hunk 6]
```diff
@@ -636,6 +636,87 @@ bool IsProcessAlive(pid_t pid) {
...
+  for (const auto &file : dir) {
+    if (!file.is_directory()) {
+      continue;
+    }
+
+    // Determine if the directory name consists of only digits (means it's a PID).
```
[the remark this answers is not in the document]
Comment (cadedaniel) [reply src/ray/util/process.cc:675]: > Is it always true that a digit only file name is the pid?

Are you asking if a pid can be non-digit? No, it is always an integer value.

> Would reading /proc/<pid>/task/<tid>/children be safer?

What is the concern? This has additional constraints; if there are multiple threads in the worker process, then we need to check each one.

File: src/ray/util/process.cc [hunk 7]
```diff
@@ -636,6 +636,87 @@ bool IsProcessAlive(pid_t pid) {
...
+std::optional<std::error_code> KillProc(pid_t pid) {
+#if defined(__linux__)
+  return {KillProcLinux(pid)};
+#else
+  return std::nullopt;
+#endif
+}
+
+#if defined(__linux__)
+static inline std::vector<pid_t> GetAllProcsWithPpidLinux(pid_t parent_pid) {
```
[the remark this answers is not in the document]
Comment (cadedaniel) [reply src/ray/util/process.cc:658]: > oh, it's a bummer that linux doesn't provide an API to return all child processes and I think this code is needed later on when we implement the raylet subreaper.

Linux does, the API is this file :) `ps` actually uses this file under the covers. Long-term, if we need more interaction with `/proc` in the Ray backend, we should build and link against [libproc-dev](https://launchpad.net/ubuntu/precise/+package/libproc-dev).

> Actually would just Process.Exec("kill $(ps -o pid= --ppid XXX)") just work

Yep -- this is the same as that. Using `Process.Exec` is a bit worse from a quality perspective as it assumes these binaries are on the PATH; in some environments like minimal container images this is not true.

File: src/ray/core_worker/core_worker.cc [hunk 8]
```diff
@@ -778,6 +829,9 @@ void CoreWorker::ForceExit(const rpc::WorkerExitType exit_type,
   RAY_LOG(WARNING) << "Force exit the process. "
                    << " Details: " << detail;
   Disconnect(exit_type, detail);
+
+  KillLeakedProcs();
```
Comment (jjyao) [inline src/ray/core_worker/core_worker.cc:833]:
```suggestion
  KillChildProcs();
```

**Review by jjyao — APPROVED** [review]
Approve since we think this is the lowest rick change for 2.4. In 2.5 we will fix the normal shutdown bug and raylet subreaper.

File: python/ray/tests/test_failure_3.py (hunk 2 above)
Comment (cadedaniel) [reply python/ray/tests/test_failure_3.py:356]: Still can't reproduce (torch shuts down successfully whenever I leak the processes). https://github.com/ray-project/ray/issues/31451#issuecomment-1463185117

Already spent a lot of time on this a few weeks ago, let's see if this fixes the issue in the user environment.

**Comment by cadedaniel** [comment]
Tests look good, the failures are unrelated. There are a few MacOS tests that still need to run but the C++ ones have already passed, which are the important ones.

Can we merge? cc @jjyao @scv119

**Comment by rkooo567** [comment]
@cadedaniel can you run release tests before merging this?

**Comment by cadedaniel** [comment]
> @cadedaniel can you run release tests before merging this?

kicked off core release tests here https://buildkite.com/ray-project/release-tests-pr/builds/34085

**Comment by cadedaniel** [comment]
Looking through release test failures (will update this as I go through them all):
* https://console.anyscale-staging.com/o/anyscale-internal/jobs/prodjob_p3u8p6fgj1cdwfecy3fhy6g76k?job-logs-tab=logs_contents_tab&pjd-section=states Unclear why it failed, a bit spooky.

**Comment by jjyao** [comment]
Failed release tests look unrelated.
